### PR TITLE
Rohit's App codebase overview

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Ignore the CI stuff. I did that to avoid running any tests for now

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/lightning_app/core/flow.py [hunk 1]
```diff
@@ -93,6 +93,7 @@ def __init__(self):
         """
         from lightning_app.runners.backends.backend import Backend
 
+        # Question: What is this state?
         self._state = set()
```
Comment (tchaton) [inline src/lightning_app/core/flow.py:97]: This is the internal variables used to know if an attribute is part of the state. We are keeping only the attribute name.

File: src/lightning_app/core/flow.py [hunk 2]
```diff
@@ -149,6 +150,8 @@ def __setattr__(self, name, value):
             elif isinstance(value, LightningWork):
                 self._works.add(name)
                 _set_child_name(self, value, name)
+                # Question: I do not understand, why are we removing it from the state
```
Comment (tchaton) [inline src/lightning_app/core/flow.py:153]: We aren't removing it from the state. We are simply naming the component as {parent_name}.{attribute_name}.

File: src/lightning_app/core/flow.py [hunk 3]
```diff
@@ -149,6 +150,8 @@ def __setattr__(self, name, value):
...
+                # and not doing that when value is a structure?
```
Comment (tchaton) [inline src/lightning_app/core/flow.py:154]: The structure is responsible for naming its own children.

File: src/lightning_app/core/flow.py [hunk 4]
```diff
@@ -180,6 +183,8 @@ def __setattr__(self, name, value):
             elif _is_json_serializable(value):
                 self._state.add(name)
 
+                # Question: Just wondering, should we even allow users to even change the values to
```
Comment (tchaton) [inline src/lightning_app/core/flow.py:186]: In theory yes, we would need to create a BaseType users can rely upon to create special behaviours with the state like Payload for example.

File: src/lightning_app/core/flow.py [hunk 5]
```diff
@@ -351,7 +355,7 @@ def schedule(
 
         .. code-block:: python
 
-            from lightning_app import LightningFlow
+            from lightning import LightningFlow
```
Comment (tchaton) [inline src/lightning_app/core/flow.py:358]: We are in the lightning_app docs.

File: src/lightning_app/utilities/proxies.py [hunk 6]
```diff
@@ -123,6 +123,7 @@ def __call__(self, *args, **kwargs):
                     # overriding the return value for next loop iteration.
                     _send_data_to_caller_queue(self.work, self.caller_queue, data, call_hash)
         if not self.parallel:
+            # Question: Didn't understand this exception. Can you improve it?
```
Comment (tchaton) [inline src/lightning_app/utilities/proxies.py:126]: Here is how it works.

When running a work with `parallel=False`, we can't simply continue the flow execution and therefore when the work hasn't completed.
In order to enable this, we raise an exception and re-run the flow from the beginning.

File: src/lightning_app/core/flow.py (hunk 2 above)
Comment (rohitgr7) [reply src/lightning_app/core/flow.py:153]: yes, that is happening inside `_set_child_name`, but I guess the name is removed from the state when an existing attribute is set to Flow/Work. But I don't think this is happening when it's set to any of the structures.

**Pull request closed without merging**